### PR TITLE
fix: Reduce memory usage for analytics service

### PR DESCRIPTION
**Pull request by jnewc**
# Description

CHKT-1963

Rework of the analytics service

1. Refactor service to simplify use of concurrency and promises
2. Ensure batches are sent once a limit is reached (currently 100 events)
3. Split service into isolated storage and reporting services

# Manual Testing

Run debug app, filter console on "Analytics", then go through drop-in or headless checkout

# Contributor Checklist

- [x]  All status checks have passed prior to code review
- [x]  I have correctly prefixed one of the following conventional commit titles:
    - [ ]  chore - no version update
    - [x]  fix - patch version update
    - [ ]  feat - minor version update
    - [ ]  BREAKING CHANGE - major version update
- [x]  I have added unit tests to a reasonable level of coverage where suitable

# Reviewer Checklist

- [ ]  I have verified that a suitable set of automated tests has been added
- [ ]  I have verified that the title prefix aligns to the code changes + whether a release is expected after merging the PR

# Other Stuff

- You can find out more about our automation checks [here](https://primerio.notion.site/iOS-Automation-Checks-198a1eb0e8994d999fb696d5902d97bb)
- Find out more about conventional commits [here](https://primerio.notion.site/Conventional-Commits-6ecfd6a0269a4db2af76d9f0537936b3)

### CODE REVIEW

File: Debug App/Tests/Unit Tests/Analytics/AnalyticsTests.swift [hunk 1]
```diff
@@ -1025,241 +582,35 @@ class AnalyticsTests: XCTestCase {
...
     }
+}
+
+
+extension Promise {
+    static func fulfilled(_ value: T) -> Promise<T> {
+        return Promise<T> { $0.fulfill(value) }
+    }
     
-    func test_race_conditions_on_creating_events_and_deleting_analytics_file() throws {
```
Comment (jnewc) [inline Debug App/Tests/Unit Tests/Analytics/AnalyticsTests.swift:1033]: This test and the one below tested concurrency by creating various types of queue and calling the `save` and `delete` functions in different combination

I've removed concurrency concerns from those methods so they now do what they say on the tin. As such, we no longer need these tests and I'll instead add more tests to exercise concurrency when calling `record` / `flush` / `sync`

File: Sources/PrimerSDK/Classes/Core/Analytics/AnalyticsStorage.swift [hunk 2]
```diff
@@ -0,0 +1,103 @@
+//
+//  AnalyticsStorage.swift
+//  PrimerSDK
+//
+//  Created by Jack Newcombe on 01/12/2023.
+//
+
+import Foundation
+
+private let analyticsURL: URL = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0].appendingPathComponent("analytics")
```
Comment (borisprimer) [inline Sources/PrimerSDK/Classes/Core/Analytics/AnalyticsStorage.swift:10]: nit-pick: Maybe change the name to indicate that this is a local storage URL... Is this fileprivate or private?

**Review by borisprimer — APPROVED** [review]
LGTM!

**Review by NQuinn27 — APPROVED** [review]
Nice work 👏

File: Sources/PrimerSDK/Classes/Core/Analytics/AnalyticsStorage.swift (hunk 2 above)
Comment (jnewc) [reply Sources/PrimerSDK/Classes/Core/Analytics/AnalyticsStorage.swift:10]: Good shout! ❤️

**Pull request merged**